### PR TITLE
Upgrade to pac4j-saml 1.8.9

### PR DESCRIPTION
**Pull request by patope**
Note!

- signed authentication request are mandatory 
- Changed default authentication binding from HTTP-Redirect to HTTP-POST. Saml specification discourages use of HTTP-Redirect and is unsupported in CloudFoundry's UAA 
- opensaml uses ServiceLoader a lot

### CODE REVIEW

**Comment by kuisathaverat** [comment]
in the roadmap is planned to upgrade to 2.0.X, I will start from this point, thx

https://issues.jenkins-ci.org/browse/JENKINS-39602

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 1]
```diff
@@ -73,6 +75,36 @@
...
+          LOG.log(Level.SEVERE, "Could not initialize opensaml service.", e);
+        }
+        return null;
+      }
+    });
+  }
+
+  // opensaml uses java.util.ServiceLoader to find service implementations. Implementations from other jars cannot
+  // be found using plugin's default context classloader.
+  private static <F, T> T runUberClassLoaderAsContextClassLoader(@Nonnull com.google.common.base.Supplier<T> runnable) {
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: I am not sure to understand the goal of this code, I think that it is not needed to load the openSAML library bundle

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 2]
```diff
@@ -374,22 +425,27 @@ private String getUsernameFromProfile(Saml2Profile saml2Profile) {
...
+  private SAML2Client newClient(StaplerRequest request, StaplerResponse response) {
     Preconditions.checkNotNull(idpMetadata);
+    final SAML2ClientConfiguration client = new SAML2ClientConfiguration();
+    client.setIdentityProviderMetadataResource(new RewriteableStringResource(idpMetadata));
+    client.setServiceProviderMetadataResource(new RewriteableStringResource());
 
-    Saml2Client client = new Saml2Client();
-    client.setIdpMetadata(idpMetadata);
-    client.setCallbackUrl(getConsumerServiceUrl());
-    client.setDestinationBindingType(SAMLConstants.SAML2_REDIRECT_BINDING_URI);
```
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: this breaks the redirect implementation and only works with post, so removes functionality

Comment (patope) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: yes, this is a bit unclear functionality as UAA/Spring saml IDP do not support redirect binding as defined by specification (http://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf  424). But, all other SAML IDPs i've seen do support it. OTOH support for redirect binding is coming to UAA too (https://github.com/cloudfoundry/uaa/pull/656). Maybe preferred binding method should be configurable?

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java (hunk 1 above)
Comment (patope) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: I don't like this either. I'm not familiar with Jenkins's classloader semantics. Is there way to make ServiceLoader to work without this?

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java (hunk 2 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java]: maybe it is better to add a field to define the binding method, to allow the user to decided.

**Comment by kuisathaverat** [comment]
finally, I upgraded to 1.9.9 https://github.com/jenkinsci/saml-plugin/pull/26

**Comment by kuisathaverat** [comment]
also, I created this [ENKINS-45857](https://issues.jenkins-ci.org/browse/JENKINS-45857) to allow to set the binding

**Pull request closed without merging**